[Quant][Inductor] Enable quantization linear pattern fusion with int8_mixed_bf16 for gelu

Le-Zheng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116004
* #114854
* #114853

**Summary**
Enable QLinear Unary pattern for gelu with int8_mix_bf16

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_int8_mixed_bf16


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Le-Zheng committed:
[Quant][Inductor] Enable quantization linear pattern fusion with int8_mixed_bf16 for gelu

[ghstack-poisoned]

leslie-fang-intel commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -136,6 +139,9 @@ def get_qlinear(users=1):
         _users=users,
     )
+    return _may_generate_pattern_with_dtype_convert(
+        computation_call, dtype=KeywordArg("to_float"), dtype_convert=is_bf16, users=2
```
I don't think we should add this in `get_qlinear`, especially the `users` are hardcoded as 2. The Linear-GeLU int8-mixed-bf16 has `to_fp32` after qlinear is because of the post op gelu, can we move it to the pattern of post op GeLU?

leslie-fang-intel replied:
I think for `none` and `tanh`, the users should be different? similar as the difference we see in the int8-mixed-fp32?

Le-Zheng replied:
Thanks for the comment! `none` and `tanh` has different `users`. I've made a change accordingly.

leslie-fang-intel reviewed (changes requested):
Please revisit the test case post op with `tanh`

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion with int8_mixed_bf16 for gelu"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion with int8_mixed_bf16 for gelu"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion with int8_mixed_bf16 for gelu"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion with int8_mixed_bf16 for gelu"


**Summary**
Enable QLinear Unary pattern for gelu with int8_mix_bf16

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_int8_mixed_bf16


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion with int8_mixed_bf16 for gelu"


**Summary**
Enable QLinear Unary pattern for gelu with int8_mix_bf16

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_int8_mixed_bf16


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion with int8_mixed_bf16 for gelu"


**Summary**
Enable QLinear Unary pattern for gelu with int8_mix_bf16

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_int8_mixed_bf16


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion with int8_mixed_bf16 for gelu"


**Summary**
Enable QLinear Unary pattern for gelu with int8_mix_bf16

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_int8_mixed_bf16


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion with int8_mixed_bf16 for gelu"


**Summary**
Enable QLinear Unary pattern for gelu with int8_mix_bf16

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_int8_mixed_bf16


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Le-Zheng committed:
Update on "[Quant][Inductor] Enable quantization linear pattern fusion with int8_mixed_bf16 for gelu"


**Summary**
Enable QLinear Unary pattern for gelu with int8_mix_bf16

**Test plan**
python test/inductor/test_mkldnn_pattern_matcher.py -k test_qlinear_gelu_int8_mixed_bf16


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
